File: fornituras-api/src/main/java/com/numobiz/solutions/fornituras/modules/transfers/service/TransferService.java
package com.numobiz.solutions.fornituras.modules.transfers.service;

import com.numobiz.solutions.fornituras.common.audit.AuditWriter;
import com.numobiz.solutions.fornituras.common.exception.BadRequestException;
import com.numobiz.solutions.fornituras.common.exception.ConflictException;
import com.numobiz.solutions.fornituras.common.exception.NotFoundException;
import com.numobiz.solutions.fornituras.modules.equipment.entity.Equipment;
import com.numobiz.solutions.fornituras.modules.equipment.entity.EquipmentStatus;
import com.numobiz.solutions.fornituras.modules.equipment.repository.EquipmentRepository;
import com.numobiz.solutions.fornituras.modules.transfers.dto.TransferCreateRequest;
import com.numobiz.solutions.fornituras.modules.transfers.dto.TransferDetail;
import com.numobiz.solutions.fornituras.modules.transfers.dto.TransferItemDetail;
import com.numobiz.solutions.fornituras.modules.transfers.dto.TransferSummary;
import com.numobiz.solutions.fornituras.modules.transfers.entity.Transfer;
import com.numobiz.solutions.fornituras.modules.transfers.entity.TransferItem;
import com.numobiz.solutions.fornituras.modules.transfers.entity.TransferStatus;
import com.numobiz.solutions.fornituras.modules.transfers.repository.TransferItemRepository;
import com.numobiz.solutions.fornituras.modules.transfers.repository.TransferRepository;
import com.numobiz.solutions.fornituras.modules.users.entity.User;
import com.numobiz.solutions.fornituras.modules.users.repository.UserRepository;
import com.numobiz.solutions.fornituras.modules.warehouses.entity.Warehouse;
import com.numobiz.solutions.fornituras.modules.warehouses.repository.WarehouseRepository;
import jakarta.persistence.criteria.Predicate;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Lógica de traslados entre almacenes. La consistencia de estado es la garantía central: crear un
 * traslado exige que cada fornitura esté <b>disponible</b> y <b>en el almacén origen</b>, y las pasa
 * a <b>en traslado</b> (bloqueadas para asignación/baja) en una sola transacción; recibir las libera
 * en el destino; cancelar las devuelve al origen. Toda mutación queda auditada (sin PII, por id).
 */
@Service
@Transactional(readOnly = true)
public class TransferService {

	private final TransferRepository repository;
	private final TransferItemRepository itemRepository;
	private final EquipmentRepository equipmentRepository;
	private final WarehouseRepository warehouseRepository;
	private final UserRepository userRepository;
	private final AuditWriter audit;

	public TransferService(
			TransferRepository repository,
			TransferItemRepository itemRepository,
			EquipmentRepository equipmentRepository,
			WarehouseRepository warehouseRepository,
			UserRepository userRepository,
			AuditWriter audit) {
		this.repository = repository;
		this.itemRepository = itemRepository;
		this.equipmentRepository = equipmentRepository;
		this.warehouseRepository = warehouseRepository;
		this.userRepository = userRepository;
		this.audit = audit;
	}

	public Page<TransferSummary> findAll(Long origenId, Long destinoId, TransferStatus status, Pageable pageable) {
		return repository.findAll(filterBy(origenId, destinoId, status), pageable).map(this::toSummary);
	}

	public TransferDetail findById(Long id) {
		return toDetail(getOrThrow(id));
	}

	@Transactional
	public TransferDetail create(TransferCreateRequest request) {
		if (request.origenId().equals(request.destinoId())) {
			throw new BadRequestException("El almacén origen y destino deben ser distintos.");
		}
		requireWarehouse(request.origenId(), "origen");
		requireActiveWarehouse(request.destinoId());

		Transfer transfer = new Transfer();
		transfer.setOrigenId(request.origenId());
		transfer.setDestinoId(request.destinoId());
		transfer.setStatus(TransferStatus.ENVIADO);
		transfer.setFechaEnvio(LocalDateTime.now());
		transfer.setCreadoPor(currentUserId());
		transfer.setObservaciones(request.observaciones());
		Transfer savedTransfer = repository.save(transfer);

		for (Long equipmentId : request.equipmentIds()) {
			Equipment equipment = requireTransferable(equipmentId, request.origenId());
			equipment.setStatus(EquipmentStatus.EN_TRASLADO);
			equipmentRepository.save(equipment);

			TransferItem item = new TransferItem();
			item.setTransferId(savedTransfer.getId());
			item.setEquipmentId(equipmentId);
			itemRepository.save(item);
		}

		audit.record("CREATE_TRANSFER", savedTransfer.getId());
		return toDetail(savedTransfer);
	}

	@Transactional
	public TransferDetail receive(Long id) {
		Transfer transfer = requireEnviado(id);
		transfer.setStatus(TransferStatus.RECIBIDO);
		transfer.setFechaRecepcion(LocalDateTime.now());
		transfer.setRecibidoPor(currentUserId());
		repository.save(transfer);

		for (TransferItem item : itemRepository.findByTransferId(id)) {
			equipmentRepository.findById(item.getEquipmentId()).ifPresent(equipment -> {
				equipment.setStatus(EquipmentStatus.DISPONIBLE);
				equipment.setWarehouseId(transfer.getDestinoId());
				equipmentRepository.save(equipment);
			});
		}

		audit.record("RECEIVE_TRANSFER", id);
		return toDetail(transfer);
	}

	@Transactional
	public TransferDetail cancel(Long id) {
		Transfer transfer = requireEnviado(id);
		transfer.setStatus(TransferStatus.CANCELADO);
		repository.save(transfer);

		// Las fornituras vuelven a disponible en el origen (no cambia su almacén).
		for (TransferItem item : itemRepository.findByTransferId(id)) {
			equipmentRepository.findById(item.getEquipmentId()).ifPresent(equipment -> {
				equipment.setStatus(EquipmentStatus.DISPONIBLE);
				equipmentRepository.save(equipment);
			});
		}

		audit.record("CANCEL_TRANSFER", id);
		return toDetail(transfer);
	}

	/** La fornitura debe existir, estar disponible y encontrarse en el almacén origen. */
	private Equipment requireTransferable(Long equipmentId, Long origenId) {
		Equipment equipment = equipmentRepository.findById(equipmentId)
				.orElseThrow(() -> new NotFoundException("Fornitura no encontrada: " + equipmentId));
		if (equipment.getStatus() != EquipmentStatus.DISPONIBLE) {
			throw new ConflictException(
					"La fornitura " + equipment.getCodigoQr() + " no está disponible (estado: " + equipment.getStatus() + ").");
		}
		if (!equipment.getWarehouseId().equals(origenId)) {
			throw new ConflictException(
					"La fornitura " + equipment.getCodigoQr() + " no está en el almacén origen.");
		}
		return equipment;
	}

	private Transfer requireEnviado(Long id) {
		Transfer transfer = getOrThrow(id);
		if (transfer.getStatus() != TransferStatus.ENVIADO) {
			throw new ConflictException("El traslado no está en curso (estado: " + transfer.getStatus() + ").");
		}
		return transfer;
	}

	private Warehouse requireWarehouse(Long id, String rol) {
		return warehouseRepository.findById(id)
				.orElseThrow(() -> new BadRequestException("Almacén " + rol + " no encontrado: " + id));
	}

	private void requireActiveWarehouse(Long id) {
		Warehouse warehouse = requireWarehouse(id, "destino");
		if (!warehouse.isActive()) {
			throw new BadRequestException("El almacén destino está inactivo.");
		}
	}

	private Transfer getOrThrow(Long id) {
		return repository.findById(id)
				.orElseThrow(() -> new NotFoundException("Traslado no encontrado: " + id));
	}

	private Long currentUserId() {
		Authentication auth = SecurityContextHolder.getContext().getAuthentication();
		if (auth == null || auth.getName() == null) {
			return null;
		}
		return userRepository.findByEmail(auth.getName()).map(User::getId).orElse(null);
	}

	private Specification<Transfer> filterBy(Long origenId, Long destinoId, TransferStatus status) {
		return (root, query, cb) -> {
			List<Predicate> predicates = new ArrayList<>();
			if (origenId != null) {
				predicates.add(cb.equal(root.get("origenId"), origenId));
			}
			if (destinoId != null) {
				predicates.add(cb.equal(root.get("destinoId"), destinoId));
			}
			if (status != null) {
				predicates.add(cb.equal(root.get("status"), status));
			}
			return cb.and(predicates.toArray(new Predicate[0]));
		};
	}

	private TransferSummary toSummary(Transfer transfer) {
		return new TransferSummary(
				transfer.getId(),
				transfer.getOrigenId(),
				warehouseName(transfer.getOrigenId()),
				transfer.getDestinoId(),
				warehouseName(transfer.getDestinoId()),
				transfer.getStatus(),
				transfer.getFechaEnvio(),
				transfer.getFechaRecepcion(),
				itemRepository.countByTransferId(transfer.getId()));
	}

	private TransferDetail toDetail(Transfer transfer) {
		List<TransferItemDetail> items = new ArrayList<>();
		for (TransferItem item : itemRepository.findByTransferId(transfer.getId())) {
			Equipment equipment = equipmentRepository.findById(item.getEquipmentId()).orElse(null);
			items.add(new TransferItemDetail(
					item.getEquipmentId(),
					equipment == null ? null : equipment.getCodigoQr(),
					equipment == null ? null : equipment.getDescripcion()));
		}
		return new TransferDetail(
				transfer.getId(),
				transfer.getOrigenId(),
				warehouseName(transfer.getOrigenId()),
				transfer.getDestinoId(),
				warehouseName(transfer.getDestinoId()),
				transfer.getStatus(),
				transfer.getFechaEnvio(),
				transfer.getFechaRecepcion(),
				transfer.getObservaciones(),
				items);
	}

	private String warehouseName(Long id) {
		return warehouseRepository.findById(id).map(Warehouse::getNombre).orElse(null);
	}
}

File: fornituras-api-dotnet/src/Fornituras.Api/Services/TransferService.cs
using Fornituras.Api.Common;
using Fornituras.Api.Data;
using Fornituras.Api.Data.Entities;
using Fornituras.Api.Dto;
using Fornituras.Api.Security;
using Microsoft.EntityFrameworkCore;
using EntityTransferStatus = Fornituras.Api.Data.Entities.TransferStatus;
using DtoTransferStatus = Fornituras.Api.Dto.TransferStatus;
using EntityEquipmentStatus = Fornituras.Api.Data.Entities.EquipmentStatus;

namespace Fornituras.Api.Services;

public sealed class TransferService(
    ApplicationDbContext db,
    CurrentUserService currentUser,
    IAuditWriter audit) : ITransferService
{
    public Task<PageResult<TransferSummary>> FindAllAsync(
        long? origenId,
        long? destinoId,
        DtoTransferStatus? status,
        PaginationQuery pagination,
        CancellationToken cancellationToken = default) =>
        FindAllInternalAsync(origenId, destinoId, status, pagination.Page, pagination.Size, cancellationToken);

    private async Task<PageResult<TransferSummary>> FindAllInternalAsync(
        long? origenId,
        long? destinoId,
        DtoTransferStatus? status,
        int page,
        int size,
        CancellationToken cancellationToken = default)
    {
        var query = db.Transfers.AsNoTracking();

        if (origenId.HasValue)
        {
            query = query.Where(t => t.OrigenId == origenId.Value);
        }

        if (destinoId.HasValue)
        {
            query = query.Where(t => t.DestinoId == destinoId.Value);
        }

        if (status.HasValue)
        {
            query = query.Where(t => t.Status == ToEntityStatus(status.Value));
        }

        query = query.OrderByDescending(t => t.FechaEnvio);
        var total = await query.CountAsync(cancellationToken);
        var transfers = await query
            .Include(t => t.Items)
            .Skip(page * size)
            .Take(size)
            .ToListAsync(cancellationToken);
        var summaries = await MapSummariesAsync(transfers, cancellationToken);
        return PageResult<TransferSummary>.From(summaries, total, page, size);
    }

    public async Task<TransferDetail> FindByIdAsync(long id, CancellationToken cancellationToken = default)
    {
        var transfer = await db.Transfers.AsNoTracking()
            .Include(t => t.Items)
            .FirstOrDefaultAsync(t => t.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Transfer not found: {id}");

        return await MapDetailAsync(transfer, cancellationToken);
    }

    public async Task<TransferDetail> CreateAsync(
        TransferCreateRequest request,
        CancellationToken cancellationToken = default)
    {
        if (request.OrigenId == request.DestinoId)
        {
            throw new BadRequestException("El almacén de origen y destino deben ser distintos.");
        }

        var origen = await db.Warehouses.FirstOrDefaultAsync(w => w.Id == request.OrigenId, cancellationToken)
            ?? throw new BadRequestException($"Almacén origen no encontrado: {request.OrigenId}");

        var destino = await db.Warehouses.FirstOrDefaultAsync(w => w.Id == request.DestinoId, cancellationToken)
            ?? throw new BadRequestException($"Almacén destino no encontrado: {request.DestinoId}");

        if (!destino.Active)
        {
            throw new BadRequestException("El almacén destino está inactivo.");
        }

        if (request.EquipmentIds.Count == 0)
        {
            throw new BadRequestException("Debe incluir al menos una fornitura.");
        }

        var equipmentList = await db.Equipment
            .Where(e => request.EquipmentIds.Contains(e.Id))
            .ToListAsync(cancellationToken);

        if (equipmentList.Count != request.EquipmentIds.Count)
        {
            throw new BadRequestException("Una o más fornituras no existen.");
        }

        foreach (var equipment in equipmentList)
        {
            if (equipment.Status != EntityEquipmentStatus.DISPONIBLE)
            {
                throw new ConflictException(
                    $"La fornitura {equipment.CodigoQr} no está disponible para traslado.");
            }

            if (equipment.WarehouseId != origen.Id)
            {
                throw new ConflictException(
                    $"La fornitura {equipment.CodigoQr} no pertenece al almacén de origen.");
            }
        }

        var userId = await currentUser.GetUserIdAsync(cancellationToken);
        var now = DateTime.UtcNow;

        var transfer = new Transfer
        {
            OrigenId = request.OrigenId,
            DestinoId = request.DestinoId,
            Status = EntityTransferStatus.ENVIADO,
            FechaEnvio = now,
            CreadoPor = userId,
            Observaciones = request.Observaciones?.Trim(),
            CreatedAt = now,
            UpdatedAt = now,
            Items = equipmentList.Select(e => new TransferItem
            {
                EquipmentId = e.Id,
                CreatedAt = now,
                UpdatedAt = now
            }).ToList()
        };

        foreach (var equipment in equipmentList)
        {
            equipment.Status = EntityEquipmentStatus.EN_TRASLADO;
            equipment.UpdatedAt = now;
        }

        db.Transfers.Add(transfer);
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("CREATE_TRANSFER", transfer.Id, cancellationToken);
        return await FindByIdAsync(transfer.Id, cancellationToken);
    }

    public async Task<TransferDetail> ReceiveAsync(long id, CancellationToken cancellationToken = default)
    {
        var transfer = await db.Transfers
            .Include(t => t.Items)
            .FirstOrDefaultAsync(t => t.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Transfer not found: {id}");

        if (transfer.Status != EntityTransferStatus.ENVIADO)
        {
            throw new ConflictException("El traslado no está en estado ENVIADO.");
        }

        var userId = await currentUser.GetUserIdAsync(cancellationToken);
        var now = DateTime.UtcNow;

        transfer.Status = EntityTransferStatus.RECIBIDO;
        transfer.FechaRecepcion = now;
        transfer.RecibidoPor = userId;
        transfer.UpdatedAt = now;

        var equipmentIds = transfer.Items.Select(i => i.EquipmentId).ToList();
        var equipmentList = await db.Equipment.Where(e => equipmentIds.Contains(e.Id)).ToListAsync(cancellationToken);

        foreach (var equipment in equipmentList)
        {
            equipment.Status = EntityEquipmentStatus.DISPONIBLE;
            equipment.WarehouseId = transfer.DestinoId;
            equipment.UpdatedAt = now;
        }

        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("RECEIVE_TRANSFER", transfer.Id, cancellationToken);
        return await FindByIdAsync(transfer.Id, cancellationToken);
    }

    public async Task<TransferDetail> CancelAsync(long id, CancellationToken cancellationToken = default)
    {
        var transfer = await db.Transfers
            .Include(t => t.Items)
            .FirstOrDefaultAsync(t => t.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Transfer not found: {id}");

        if (transfer.Status != EntityTransferStatus.ENVIADO)
        {
            throw new ConflictException("El traslado no está en estado ENVIADO.");
        }

        var now = DateTime.UtcNow;
        transfer.Status = EntityTransferStatus.CANCELADO;
        transfer.UpdatedAt = now;

        var equipmentIds = transfer.Items.Select(i => i.EquipmentId).ToList();
        var equipmentList = await db.Equipment.Where(e => equipmentIds.Contains(e.Id)).ToListAsync(cancellationToken);

        foreach (var equipment in equipmentList)
        {
            equipment.Status = EntityEquipmentStatus.DISPONIBLE;
            equipment.UpdatedAt = now;
        }

        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("CANCEL_TRANSFER", transfer.Id, cancellationToken);
        return await FindByIdAsync(transfer.Id, cancellationToken);
    }

    private async Task<IReadOnlyList<TransferSummary>> MapSummariesAsync(
        IReadOnlyList<Transfer> transfers,
        CancellationToken cancellationToken)
    {
        var warehouseIds = transfers.SelectMany(t => new[] { t.OrigenId, t.DestinoId }).Distinct();
        var names = await db.Warehouses.AsNoTracking()
            .Where(w => warehouseIds.Contains(w.Id))
            .ToDictionaryAsync(w => w.Id, w => w.Nombre, cancellationToken);

        return transfers.Select(t => new TransferSummary(
            t.Id,
            t.OrigenId,
            names.GetValueOrDefault(t.OrigenId, "?"),
            t.DestinoId,
            names.GetValueOrDefault(t.DestinoId, "?"),
            ToDtoStatus(t.Status),
            t.FechaEnvio,
            t.FechaRecepcion,
            t.Items?.Count ?? 0)).ToList();
    }

    private async Task<TransferDetail> MapDetailAsync(Transfer transfer, CancellationToken cancellationToken)
    {
        var origen = await db.Warehouses.AsNoTracking()
            .Where(w => w.Id == transfer.OrigenId)
            .Select(w => w.Nombre)
            .FirstAsync(cancellationToken);
        var destino = await db.Warehouses.AsNoTracking()
            .Where(w => w.Id == transfer.DestinoId)
            .Select(w => w.Nombre)
            .FirstAsync(cancellationToken);

        var equipmentIds = transfer.Items.Select(i => i.EquipmentId).ToList();
        var equipmentMap = await db.Equipment.AsNoTracking()
            .Where(e => equipmentIds.Contains(e.Id))
            .ToDictionaryAsync(e => e.Id, cancellationToken);

        var items = transfer.Items.Select(i =>
        {
            equipmentMap.TryGetValue(i.EquipmentId, out var eq);
            return new TransferItemDetail(i.EquipmentId, eq?.CodigoQr ?? "?", eq?.Descripcion);
        }).ToList();

        return new TransferDetail(
            transfer.Id,
            transfer.OrigenId,
            origen,
            transfer.DestinoId,
            destino,
            ToDtoStatus(transfer.Status),
            transfer.FechaEnvio,
            transfer.FechaRecepcion,
            transfer.Observaciones,
            items);
    }

    private static EntityTransferStatus ToEntityStatus(DtoTransferStatus status) =>
        Enum.Parse<EntityTransferStatus>(status.ToString());

    private static DtoTransferStatus ToDtoStatus(EntityTransferStatus status) =>
        Enum.Parse<DtoTransferStatus>(status.ToString());
}
